[AutoDiff] Add SIL differentiability witnesses.

dan-zheng opened the pull request:
SIL differentiability witnesses are a new top-level SIL construct mapping "original" SIL functions to derivatives. They will replace SIL function `[differentiable]` attributes, additionally enabling cross-module retroactive derivative registration.

SIL differentiability witnesses have the following components:
* Original `SILFunction`.
* Linkage.
* Parameter indices (`IndexSubset`).
* Result indices (`IndexSubset`).
* Derivative generic signature (optional).
* JVP function (optional).
* VJP function (optional).

Example syntax:

```
sil_differentiability_witness hidden [parameters 0 1] [results 0] [where τ_0_0 : _Differentiable] @foo : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0, Float) -> @out τ_0_0 {
  jvp: @AD__foo__jvp_src_0_wrt_0 : $@convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0, Float) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector, Float) -> @out τ_0_0.TangentVector)
  vjp: @AD__foo__vjp_src_0_wrt_0 : $@convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0, Float) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector) -> (@out τ_0_0.TangentVector, Float))
}
```

---

This patch adds the `SILDifferentiabilityWitness` data structure, along with parsing, printing, verification, and serialization (including lookup by key).

The master issue [TF-866](https://bugs.swift.org/browse/TF-866) tracks follow-up, including SILGen/IRGen/differentiation transform changes.

---

[Early forum discussion.](https://forums.swift.org/t/help-needed-with-retroactive-differentiability/19927)

rxwei committed:
Diff witness

dan-zheng committed:
Rename SILDifferentiabilityWitness.h.

dan-zheng committed:
Update SILDifferentiabilityWitness definition.

dan-zheng committed:
Add SILDifferentiabilityWitness to SILModule.

dan-zheng committed:
[WIP] Start SILDifferentiabilityWitness parsing/printing.

- Printing compiles but may require changes to be parseable. Namely,
  if there is no suitable utility for parsing standalone generic
  signatures, changes are needed.
- Parsing is a stub.
- Note: it is difficult to test parsing/printing without generating
  SILDifferentiabilityWitness instances.

dan-zheng committed:
Use improved syntax.

`parameters (0, 1, ...) results (0, 1, ...) where <...>`

dan-zheng committed:
Merge branch 'tensorflow' of github.com:apple/swift into sil-differentiability-witness

dan-zheng committed:
Finish parsing/printing/serialization.

dan-zheng committed:
Revamp serialization to enable lookup by key.

Use mangling to support string key lookup.

dan-zheng committed:
Add SIL verification.

dan-zheng committed:
Add miscellaneous todo comments.

dan-zheng committed:
Add round-trip parsing/printing test.

dan-zheng committed:
Clean up.

dan-zheng committed:
Add Swift source for parsing test.

dan-zheng committed:
Merge branch 'tensorflow' of github.com:apple/swift into sil-differentiability-witness

dan-zheng committed:
`AutoDiffIndexSubset` -> `IndexSubset`

rxwei reviewed (commented):
1. Because type signatures are long and sparse, I'd suggest bringing the configuration to the front of the declaration and using square brackets for them, for example:

    ```
    sil_differentiability_witness hidden [parameters 0 1] [results 0] [where τ_0_0 : _Differentiable] @foo : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0, Float) -> @out τ_0_0 {
      jvp: @AD__foo__jvp_src_0_wrt_0 : $@convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0, Float) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector, Float) -> @out τ_0_0.TangentVector)
      vjp: @AD__foo__vjp_src_0_wrt_0 : $@convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0, Float) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector) -> (@out τ_0_0.TangentVector, Float))
    }
    ```
2. Parts of what `SILDifferentiabilityWitnessKey` stores could have been made a more general concept: `AutoDiffConfig`. How about defining a `AutoDiffConfig` in AutoDiff.h to store parameter indices, result indices, and the derivative generic signature? Then `SILDifferentiabilityWitnessKey` can just be defined as `std::pair<StringRef, AutoDiffConfig>`.

3. When the function name of a `sil_differentiability_witness` is demangleable, consider printing a descriptive comment above the witness declaration, for example:

    ```console
    // differentiability witness for closure #1 in foo<A>(_:_:) 
    sil_differentiability_witness hidden [parameters 0 1] [results 0] [where τ_0_0 : _Differentiable] @someDemangleableName : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0, Float) -> @out τ_0_0 {
    ```

dan-zheng committed:
Minor fix.

rxwei commented on include/swift/SIL/SILDifferentiabilityWitness.h:
```diff
@@ -0,0 +1,132 @@
+      public SILAllocated<SILDifferentiabilityWitness>
+{
+private:
+  /// The module which contains the differentiability witness.
```
You also need to store a source location here for differentiation diagnostics.

dan-zheng replied:
Chatted in person, storing source location may not make the most sense.
Added `DeclAttribute *` field (representing AST `@differentiable` or `@differentiating attribute) in https://github.com/apple/swift/pull/27487/commits/4ddcd3f059b686dcc6137bf4a7455419b94798aa.

dan-zheng committed:
Update differentiability witness syntax.

Print original function name in comment.

```
// differentiability witness for foo
sil_differentiability_witness hidden [parameters 0 1] [results 0] [where τ_0_0 : _Differentiable] @foo
```

dan-zheng committed:
Add `AutoDiffConfig` and use in `SILDifferentiabilityWitnessKey`.

rxwei commented on include/swift/AST/AutoDiff.h:
```diff
@@ -208,6 +208,41 @@ struct AutoDiffDerivativeFunctionKind {
+/// - Derivative generic signature (optional).
+// TODO(TF-893): Use `AutoDiffConfig` in `AutoDiffDerivativeFunctionIdentifier`
+// to avoid duplication.
+class AutoDiffConfig : public llvm::FoldingSetNode {
```
* `AutoDiffConfig` does not need to be uniqued. It is only 3 words long and can simply be a POD.
* Fields do not need to be `const`.
* No getters are necessary.

```swift
struct AutoDiffConfig {
  IndexSubset *parameterIndices;
  IndexSubset *resultIndices;
  GenericSignature *derivativeGenericSignature;
};
```

dan-zheng replied:
Chatted in person; POD makes sense because all three fields are already individually uniqued. `AutoDiffDerivativeFunctionIdentifier` is not necessarily a good example and may be removed later when VJPs are removed and only JVPs exist.

Done in https://github.com/apple/swift/pull/27487/commits/10700e714f16dfb35ff8930e55493976f0a829f2.

dan-zheng committed:
Change `AutoDiffConfig` to a POD.

The contents of `AutoDiffConfig` are all uniqued, so uniquing the product
does not make sense.

dan-zheng committed:
Add `DeclAttribute *` to `SILDifferentiabilityWitness`.

Unserialized, to be used for diagnostics.
Will revisit later when revamping the differentiation transform.

dan-zheng committed:
Clean up.

rxwei commented on include/swift/AST/AutoDiff.h:
```diff
@@ -363,10 +377,44 @@ class VectorSpace {
+  static AutoDiffConfig getTombstoneKey() {
+    auto *ptr = llvm::DenseMapInfo<void *>::getTombstoneKey();
+    return {static_cast<IndexSubset *>(ptr),
+        static_cast<IndexSubset *>(ptr),
```
Fix indentation.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/771143ae652d5b9cc49fc152863b4d133ee9c9c2.

rxwei commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -686,6 +686,18 @@ ERROR(sil_witness_assoc_conf_not_found,none,
+ERROR(sil_diff_witness_expected_keyword,PointsToFirstBadToken,
+      "expected '%0' in differentiability witness", (StringRef))
+ERROR(sil_diff_witness_expected_parameter_list,PointsToFirstBadToken,
+      "expected an comma-separated list of parameter indices, e.g. (0, 1)", ())
```
The syntax is no longer comma-separated or parenthesized.

dan-zheng replied:
Fixed in https://github.com/apple/swift/pull/27487/commits/771143ae652d5b9cc49fc152863b4d133ee9c9c2:
```
expected a space-separated list of indices, e.g. '0 1'
```

rxwei commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -686,6 +686,18 @@ ERROR(sil_witness_assoc_conf_not_found,none,
+      "expected ']' to end 'differentiable' attribute", ())
+ERROR(sil_diff_witness_expected_parameter_index,PointsToFirstBadToken,
+      "expected the index of a parameter to differentiate w.r.t.", ())
+ERROR(sil_diff_witness_expected_source_index,PointsToFirstBadToken,
```
The word "source" has been a bad word in the implementation for a long time ([SR-8888](https://bugs.swift.org/browse/SR-8888)) -- please avoid using it. "Differentiate from" is also bad. We should use "result index" and "differentiate with respect to".

dan-zheng replied:
Copied code, sorry. Updated in https://github.com/apple/swift/pull/27487/commits/771143ae652d5b9cc49fc152863b4d133ee9c9c2.

rxwei commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -686,6 +686,18 @@ ERROR(sil_witness_assoc_conf_not_found,none,
+ERROR(sil_diff_witness_expected_rsquare,PointsToFirstBadToken,
+      "expected ']' to end 'differentiable' attribute", ())
+ERROR(sil_diff_witness_expected_parameter_index,PointsToFirstBadToken,
+      "expected the index of a parameter to differentiate w.r.t.", ())
```
```suggestion
      "expected the index of a parameter to differentiate with respect to", ())
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/771143ae652d5b9cc49fc152863b4d133ee9c9c2.

rxwei commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -686,6 +686,18 @@ ERROR(sil_witness_assoc_conf_not_found,none,
+ERROR(sil_diff_witness_expected_parameter_list,PointsToFirstBadToken,
+      "expected an comma-separated list of parameter indices, e.g. (0, 1)", ())
+ERROR(sil_diff_witness_expected_rsquare,PointsToFirstBadToken,
+      "expected ']' to end 'differentiable' attribute", ())
```
Bad wording. Looks like it's copied from `[differentiable]` parsing.

dan-zheng replied:
That diagnostic is actually not used. Removed in https://github.com/apple/swift/pull/27487/commits/771143ae652d5b9cc49fc152863b4d133ee9c9c2.

dan-zheng committed:
Address review feedback.

rxwei commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -686,6 +686,16 @@ ERROR(sil_witness_assoc_conf_not_found,none,
+ERROR(sil_diff_witness_expected_index_list,PointsToFirstBadToken,
+      "expected a space-separated list of indices, e.g. '0 1'", ())
+ERROR(sil_diff_witness_expected_parameter_index,PointsToFirstBadToken,
+      "expected a parameter index to differentiate with respect to.", ())
```
Remove periods in diagnostics
```suggestion
      "expected a parameter index to differentiate with respect to", ())
```

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -6742,6 +6745,242 @@ bool SILParserTUState::parseSILDefaultWitnessTable(Parser &P) {
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+// TODO(TF-893): Dedupe with `SILParser::convertRequirements` upstream.
```
Could you please add a comment to explain why this duplication is necessary?

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

`convertRequirements` is currently defined on `SILParser`, but it needs to be moved to `Parser` for differentiability witness parsing.

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -6742,6 +6745,242 @@ bool SILParserTUState::parseSILDefaultWitnessTable(Parser &P) {
+  // Parse differentiability witness body.
+  SILFunction *jvp = nullptr;
+  SILFunction *vjp = nullptr;
+  if (P.Tok.is(tok::l_brace)) {
```
```suggestion
  SourceLoc lBraceLoc;
  if (P.consumeIf(Tok::l_brace, lBraceLoc)) {
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -686,6 +686,16 @@ ERROR(sil_witness_assoc_conf_not_found,none,
       "sil protocol conformance not found", ())
 
+// SIL differentiability witnesses
+ERROR(sil_diff_witness_expected_keyword,PointsToFirstBadToken,
```
Keyword has a specific meaning in SIL syntax. Use "token" instead.
```suggestion
ERROR(sil_diff_witness_expected_token,PointsToFirstBadToken,
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on include/swift/SIL/SILDifferentiabilityWitness.h:
```diff
@@ -0,0 +1,156 @@
+  static AutoDiffConfig *
+  getAutoDiffConfig(SILModule &module, IndexSubset *parameterIndices,
+                    IndexSubset *resultIndices,
+                    GenericSignature *derivativeGenSig);
```
Remove dead code.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on include/swift/SIL/SILDifferentiabilityWitness.h:
```diff
@@ -0,0 +1,156 @@
+  GenericSignature *getDerivativeGenericSignature() const {
+    return derivativeGenericSignature;
+  }
+  SILFunction *getJVP() const { return jvp; }
```
It seems a bit odd to not have `setJVP` and `setVJP` when there's `setDerivative`.

dan-zheng replied:
Sorry, `setDerivative` isn't used in this PR but is used in my next patch. It's more convenient for callers than `setJVP` and `setVJP`, since callers have to do a `switch` over derivative kind.

I'll remove the utility from this patch since it's not yet used.

dan-zheng replied:
Actually, I decided to add `setJVP` and `setVJP`. Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on include/swift/SIL/SILDifferentiabilityWitness.h:
```diff
@@ -0,0 +1,156 @@
+  DeclAttribute *getAttribute() const { return attribute; }
+
+  /// Verify that the differentiability witness is well-formed.
+  void verify(const SILModule &M) const;
```
Stick to one naming convention. I see that you are using mostly camel case in this file, so it's better to make that consistent. Same for `print`'s arguments below.

```suggestion
  void verify(const SILModule &module) const;
```

dan-zheng replied:
Fair. I will choose not to update the variable names in the defining files (`SILVerifier.cpp`, `SILPrinter.cpp`) to match the local formatting of those files.

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -6742,6 +6745,242 @@ bool SILParserTUState::parseSILDefaultWitnessTable(Parser &P) {
+      if (P.parseToken(tok::colon, diag::sil_diff_witness_expected_keyword,
+                       ":"))
+        return true;
+      Scope Body(&P, ScopeKind::FunctionBody);
```
I'd suggest sticking to one naming convention throughout one piece of code.
```suggestion
      Scope body(&P, ScopeKind::FunctionBody);
```

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -6742,6 +6745,242 @@ bool SILParserTUState::parseSILDefaultWitnessTable(Parser &P) {
+      return true;
+  }
+
+  // TODO(TF-893): Parse `isSerialized` flag.
```
Is it just `sil_differentiability_witness hidden [serialized] ...`? It should be trivial to tackle within this patch.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/27d7abc5cb7a9456b93a9f934a993aa0c26da5d0.

Manually verified printing/parsing, I chose not to add an additional test to keep the test small.

rxwei commented on lib/SIL/SILDifferentiabilityWitness.cpp:
```diff
@@ -0,0 +1,42 @@
+    IndexSubset *parameterIndices, IndexSubset *resultIndices,
+    GenericSignature *derivativeGenSig, SILFunction *jvp, SILFunction *vjp,
+    bool isSerialized, DeclAttribute *attribute) {
+  void *buf = module.allocate(sizeof(SILDifferentiabilityWitness),
```
`SILDifferentiabilityWitness` does not store trailing objects, so you don't need to call `SILModule::allocate` or calculate the size manually. See suggestion below.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/SIL/SILDifferentiabilityWitness.cpp:
```diff
@@ -0,0 +1,42 @@
+    bool isSerialized, DeclAttribute *attribute) {
+  void *buf = module.allocate(sizeof(SILDifferentiabilityWitness),
+                              alignof(SILDifferentiabilityWitness));
+  auto *diffWitness = ::new (buf) SILDifferentiabilityWitness(
```
```suggestion
  auto *diffWitness = new (module) SILDifferentiabilityWitness(
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3026,6 +3057,72 @@ void SILDefaultWitnessTable::dump() const {
+  OS << "[parameters ";
+  interleave(getParameterIndices()->getIndices(),
+             [&](unsigned index) { OS << index; },
+             [&] { OS << " "; });
```
```suggestion
             [&] { OS << ' '; });
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3026,6 +3057,72 @@ void SILDefaultWitnessTable::dump() const {
+void SILDifferentiabilityWitness::print(
+    llvm::raw_ostream &OS, bool verbose) const {
+  OS << "// differentiability witness for "
+     << demangleSymbol(originalFunction->getName()) << "\n";
```
Use a character literal when outputting one character. 
```suggestion
     << demangleSymbol(originalFunction->getName()) << '\n';
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3026,6 +3057,72 @@ void SILDefaultWitnessTable::dump() const {
+  OS << "] [results ";
+  interleave(getResultIndices()->getIndices(),
+             [&](unsigned index) { OS << index; },
+             [&] { OS << " "; });
```
```suggestion
             [&] { OS << ' '; });
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3026,6 +3057,72 @@ void SILDefaultWitnessTable::dump() const {
+    }
+    if (!requirements.empty()) {
+      OS << " [where ";
+      auto SubPrinter = PrintOptions::printSIL();
```
```suggestion
      auto subPrinter = PrintOptions::printSIL();
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3026,6 +3057,72 @@ void SILDefaultWitnessTable::dump() const {
+  // }
+  OS << " {\n";
+  if (jvp)
+    OS << "  jvp: @" << jvp->getName() << " : " << jvp->getLoweredType() << "\n";
```
```suggestion
    OS << "  jvp: @" << jvp->getName() << " : " << jvp->getLoweredType() << '\n';
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3026,6 +3057,72 @@ void SILDefaultWitnessTable::dump() const {
+  if (jvp)
+    OS << "  jvp: @" << jvp->getName() << " : " << jvp->getLoweredType() << "\n";
+  if (vjp)
+    OS << "  vjp: @" << vjp->getName() << " : " << vjp->getLoweredType() << "\n";
```
```suggestion
    OS << "  vjp: @" << vjp->getName() << " : " << vjp->getLoweredType() << '\n';
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3026,6 +3057,72 @@ void SILDefaultWitnessTable::dump() const {
+    llvm::raw_ostream &OS, bool verbose) const {
+  OS << "// differentiability witness for "
+     << demangleSymbol(originalFunction->getName()) << "\n";
+  // sil_differentiability_witness @original-function-name : $original-sil-type
```
Correct the syntax.
```suggestion
  // sil_differentiability_witness [parameters ...] [results ...] [where ...] @original-function-name : $original-sil-type
```

dan-zheng replied:
The suggested comment is a bit too long. Currently, each printed section has its own comment, which I think is a bit clearer, if that's okay.

Clarified comments here to be a bit more clear.

rxwei commented on lib/Serialization/DeserializeSIL.cpp:
```diff
@@ -3280,6 +3296,90 @@ SILDeserializer::lookupDefaultWitnessTable(SILDefaultWitnessTable *existingWt) {
+  if (iter == DifferentiabilityWitnessList->end())
+    return nullptr;
+
+  auto *diffWitness = readDifferentiabilityWitness(*iter);
```
Return this directly?

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -2273,11 +2285,21 @@ void SILSerializer::writeIndexTables() {
     Offset.emit(ScratchRecord, sil_index_block::SIL_PROPERTY_OFFSETS,
                 PropertyOffset);
   }
-  
```
revert this?

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -2469,6 +2491,54 @@ writeSILDefaultWitnessTable(const SILDefaultWitnessTable &wt) {
+    vjpID = S.addUniquedStringRef(vjp->getName());
+  }
+  SmallVector<unsigned, 8> parameterAndResultIndices(
+       dw.getParameterIndices()->begin(), dw.getParameterIndices()->end());
```
```suggestion
      dw.getParameterIndices()->begin(), dw.getParameterIndices()->end());
```

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27487/commits/2073f863826fb305c1fc105a520d84de1e1463ec.

rxwei commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -2469,6 +2491,54 @@ writeSILDefaultWitnessTable(const SILDefaultWitnessTable &wt) {
+  size_t nameLength = mangledKey.size();
+  char *stringStorage =
+      static_cast<char *>(StringTable.Allocate(nameLength, 1));
+  std::memcpy(stringStorage, mangledKey.data(), nameLength);
```
Use `std::uninitialized_copy` instead?

dan-zheng replied:
I get some errors using `std::uninitialized_copy`:
```
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/stl_construct.h:75:13: error: static_cast from 'const char *' to 'void *' is not allowed
    { ::new(static_cast<void*>(__p)) _T1(std::forward<_Args>(__args)...); }
            ^~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/stl_uninitialized.h:83:8: note: in instantiation of function template specialization 'std::_Construct<const char, char &>' requested here
                std::_Construct(std::__addressof(*__cur), *__first);
                     ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.4.0/../../../../include/c++/7.4.0/bits/stl_uninitialized.h:134:2: note: in instantiation of function template specialization 'std::__uninitialized_copy<false>::__uninit_copy<char *, const char *>' requested here
        __uninit_copy(__first, __last, __result);
        ^
/home/danielzheng/swift-merge/swift/lib/Serialization/SerializeSIL.cpp:2503:8: note: in instantiation of function template specialization 'std::uninitialized_copy<char *, const char *>' requested here
  std::uninitialized_copy(stringStorage, stringStorage + nameLength,
       ^
1 error generated.
```

I copied this logic including `std::memcpy` from existing code. I'd prefer to defer this minor improvement to speed up progress, existing code can also be updated.

dan-zheng committed:
Address review feedback.

dan-zheng committed:
Parse/print `[serialized]` flag.

Manually verified parsing/printing.
Chose not to add additional test for now to keep the test small.

rxwei reviewed (commented):
Are there serialization tests?

rxwei reviewed (approved):
The implementation LGTM, but I just wanted to make sure serialization is tested.

rxwei commented on test/AutoDiff/sil_differentiability_witness_parse.sil:
```diff
@@ -0,0 +1,145 @@
+// static AdditiveArithmetic<>.zero.getter
+sil [serialized] [always_inline] @$ss18AdditiveArithmeticPss27ExpressibleByIntegerLiteralRzrlE4zeroxvgZ : $@convention(method) <τ_0_0 where τ_0_0 : AdditiveArithmetic, τ_0_0 : ExpressibleByIntegerLiteral> (@thick τ_0_0.Type) -> @out τ_0_0
+
+sil_differentiability_witness hidden [parameters 0 1] [results 0] [where τ_0_0 : _Differentiable] @foo : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0, Float) -> @out τ_0_0 {
```
Also, this test file doesn't have to be this complicated, since it's just testing round-trip parsing. In particular, none of the function bodies above needed to exist in order for this test to pass. Simplifying this test file will also allow you to write more of actual, effective tests.

dan-zheng replied:
Thanks - simplified parsing/printing tests in https://github.com/apple/swift/pull/27487/commits/ef5f53cf32237810172e3aa5d6233fac2ddf975b.

There are two tests now:
```
// Test public non-generic function.
// SIL differentiability witness:
// - Has public linkage (implicit).
// - Has no `where` clause.

// Test internal generic function.
// SIL differentiability witness:
// - Has hidden linkage.
// - Has `where` clause.
```

dan-zheng replied:
Added serialization/deserialization test in https://github.com/apple/swift/pull/27487/commits/df9cd497902873bb95b92fca184f51564c16ab7d.

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -6742,6 +6745,250 @@ bool SILParserTUState::parseSILDefaultWitnessTable(Parser &P) {
+}
+
+/// decl-sil-differentiability-witness ::=
+///   'sil_differentiability_witness'
```
Missing linkage here.

dan-zheng replied:
Thanks, fixed in https://github.com/apple/swift/pull/27487/commits/ef5f53cf32237810172e3aa5d6233fac2ddf975b.

dan-zheng committed:
Add parsing/printing tests, address review feedback.

dan-zheng commented on lib/SIL/SILPrinter.cpp:
```diff
@@ -3026,6 +3057,75 @@ void SILDefaultWitnessTable::dump() const {
+      interleave(requirements,
+                 [&](Requirement req) {
+                   req.print(OS, subPrinter);
+                   return;
```
Done in https://github.com/apple/swift/commit/ef5f53cf32237810172e3aa5d6233fac2ddf975b.

rxwei commented on test/AutoDiff/sil_differentiability_witness_parse.sil:
```diff
@@ -0,0 +1,73 @@
+}
+
+sil @AD__foo__jvp_src_0_wrt_0 : $@convention(thin) (Float) -> (Float, @owned @callee_guaranteed (Float) -> Float) {
+bb0(%0 : $Float):
```
Even `undef` can be removed! You should be able to remove the entirety of the body (from `{` through `}`).

rxwei replied:
Same for other two functions that currently have `undef`.

dan-zheng replied:
Note: parsing/printing worked for bodiless "externally available" function declarations but not internal ones:

```
SIL verification failed: external declaration of internal SILFunction not allowed: F->isAvailableExternally()
In function:
// generic
sil hidden @generic : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0, Float) -> @out τ_0_0
```

Deserialization, however, did not work for bodiless "externally available" functions. More info coming.

dan-zheng replied:
Differentiability witness deserialization does not work for bodiless functions, regardless of linkage (`public`, `public_external`, etc): `SILDeserializer::getFuncForReference(StringRef name)` returns null for bodiless original/JVP/VJP functions.

Perhaps this is due to the test `RUN` lines (i.e. round-tripping `sil-opt` on the output of `-emit-sib`).

---

I found that [vtable entry](https://github.com/apple/swift/blob/65ebee0c7cea75ed757e179b0a9898b0cb4ecde7/lib/Serialization/DeserializeSIL.cpp#L2857) and [witness table entry](https://github.com/apple/swift/blob/65ebee0c7cea75ed757e179b0a9898b0cb4ecde7/lib/Serialization/DeserializeSIL.cpp#L2996) deserialization check whether the output of `getFuncForReference(StringRef)` is null.

I also found a `SILDeserializer::getFuncForReference(StringRef name, SILType type)` overload, which always returns a `SILFunction *`, creating a declaration if it does not exist. This is used by deserialization for various instructions like `function_ref`. I'm not sure using this is correct, but we could use it if we also serialized original/JVP/VJP function type.

---

I'll defer investigation until it's deemed necessary.

rxwei replied:
Not supporting bodiless functions is not quite right. We should be able to use any function declaration in differentiability witnesses.

dan-zheng committed:
Fix serialization and add test.

Note: deserialization does not work when SIL differentiability witness
references bodyless function declarations.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented on lib/SIL/SILVerifier.cpp:
```diff
@@ -5350,6 +5350,44 @@ void SILGlobalVariable::verify() const {
+        getParameterIndices(), /*resultIndex*/ *getResultIndices()->begin(),
+        AutoDiffDerivativeFunctionKind::JVP, M.Types,
+        LookUpConformanceInModule(M.getSwiftModule()), derivativeCanGenSig);
+    SILVerifier(*jvp).requireSameType(
```
We want a same type assertion like `SILVerifier::requireSameType`, but using `SILVerifier` is not right because it prints the message on the given `SILFunction`:

```
SIL verification failed: VJP type does not match expected VJP type
  $@convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector) -> @out Float)
  $@convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector) -> @out τ_0_0.TangentVector)
In function:
// AD__foo__vjp_src_0_wrt_0
sil private @AD__foo__vjp_src_0_wrt_0 : $@convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector) -> @out Float)
```

We want to print the message on the differentiability witness instead.

dan-zheng replied:
Fixed in https://github.com/apple/swift/pull/27487/commits/d673b7056cd03dbdfb77c05a23ac0ad09a433315:
```
SIL verification failed: VJP type does not match expected VJP type
  @convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector) -> @out Float)
  @convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector) -> @out τ_0_0.TangentVector)

In differentiability witness:
// differentiability witness for foo
sil_differentiability_witness hidden [parameters 0] [results 0] [where τ_0_0 : _Differentiable] @foo : $@convention(thin) <τ_0_0> (@in_guaranteed τ_0_0) -> @out τ_0_0 {
  jvp: @AD__foo__jvp_src_0_wrt_0 : $@convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector) -> @out τ_0_0.TangentVector)
  vjp: @AD__foo__vjp_src_0_wrt_0 : $@convention(thin) <τ_0_0 where τ_0_0 : _Differentiable> (@in_guaranteed τ_0_0) -> (@out τ_0_0, @owned @callee_guaranteed (@in_guaranteed τ_0_0.TangentVector) -> @out Float)
}
```

dan-zheng committed:
Fix verification.

Verification assertion messages should appear on the differentiability witness.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
